### ISSUE TITLE
Step 2 error: error in run_cutadapt(): KeyError('3')

### ISSUE THREAD

**Issue by laninsky** [dereneaton/ipyrad#289]
Hi folks, I'm running into a problem on Step 2 with one of my single-end GBS samples - a sample represented by two technical replicates that have different barcodes. I've set it (sample "3") up in the barcodes file as:
```
3       TATCTAG
3       TTCGCACA
```
Things seem to go OK through the first step, combining these replicates for that sample as desired. 3_R1_.fastq.gz is located in the lamprey_fastqs folder and about twice the size of the other samples (as you might expect from having it be represented by two replicates):
```
  Warning: technical replicates (same name) will be combined.
  New Assembly: lamprey
  Warning: technical replicates (same name) will be combined.
  host compute node: [7 cores] on compute-e1-077-p

  Step 1: Demultiplexing fastq data to Samples

 ```
However, although the rest of the samples finish through all the steps OK, sample 3 can't make it through Step 2. 
```
 Step 2: Filtering reads 

  found an error in step2; see ipyrad_log.txt

  Step 3: Clustering/Mapping reads
    Sample not ready for clustering. First run step2 on sample: 3
```
I went back and branched the assembly to try and pull this guy through (also had problems with slow spin-up times on this cluster hence the ipcluster stuff). I got the same error referenced in the STDOUT above as I did when I reran it using the following:
```
export PATH="/projects/uoo00110/bin/miniconda2/bin:$PATH"
export HOME="/projects/uoo00110"
ipcluster start --n 16 --daemonize
sleep 240
ipyrad -p params-lamprey_3_only.txt -s 2 -f -d --ipcluster
```
From the standard out: 
```  
found an error in step2; see ipyrad_log.txt
INFO:ipyrad:debugging turned off
```
From the log:
```
ipyrad [v.0.7.22]
  Interactive assembly and analysis of RAD-seq data
 -------------------------------------------------------------
  Begin run: 2018-03-22 11:52
  Using args {'preview': False, 'force': True, 'threads': 2, 'results': False, 'quiet': False, 'merge': None, 'ipcluster': 'default', 'cores': 0, 'params': 'params-lamprey_3_only.txt', 'branch': None, 'steps': '2', 'debug': True, 'new': None, 'download': None, 'MPI': False}
  Platform info: ('Linux', 'compute-b1-043-p', '2.6.32-279.14.1.el6.x86_64', '#1 SMP Mon Oct 15 13:44:51 EDT 2012', 'x86_64')2018-03-22 11:52:04,101    pid=26615 
      [load.py]       DEBUG   skipping: no svd results present in old assembly
2018-03-22 11:52:12,796         pid=26615       [rawedit.py]    INFO    [7239922.0]
2018-03-22 11:52:19,104         pid=26615       [rawedit.py]    ERROR   error in run_cutadapt(): KeyError('3')
2018-03-22 11:52:21,181         pid=26615       [__init__.py]   INFO    debugging turned off
```
Given all the other samples processed just fine (and the replicates processed fine when I coded them as separate samples), I strongly suspect the issue might be that I'm trying to merge two technical replicates with different barcodes/different barcode lengths. Any help or ideas where I might have stuffed up would be appreciated!

**Comment by laninsky** [comment]
params file just in case it is useful:
```
------- ipyrad params file (v.0.7.22)-------------------------------------------
lamprey_3_only                 ## [0] [assembly_name]: Assembly name. Used to name output directories for assembly steps
/gpfs1m/projects/uoo00110      ## [1] [project_dir]: Project dir (made in curdir if not present)
/gpfs1m/projects/uoo00110/SQ0419_CA5EWANXX_s_4_fastq.txt.gz ## [2] [raw_fastq_path]: Location of raw non-demultiplexed fastq files
/gpfs1m/projects/uoo00110/barcodes_3merged.txt ## [3] [barcodes_path]: Location of barcodes file
/projects/uoo00110/lamprey_fastqs/*.fastq.gz                               ## [4] [sorted_fastq_path]: Location of demultiplexed/sorted fastq files
denovo                         ## [5] [assembly_method]: Assembly method (denovo, reference, denovo+reference, denovo-reference)
                               ## [6] [reference_sequence]: Location of reference sequence file
gbs                            ## [7] [datatype]: Datatype (see docs): rad, gbs, ddrad, etc.
TGCAG, TGCAG                   ## [8] [restriction_overhang]: Restriction overhang (cut1,) or (cut1, cut2)
5                              ## [9] [max_low_qual_bases]: Max low quality base calls (Q<20) in a read
33                             ## [10] [phred_Qscore_offset]: phred Q score offset (33 is default and very standard)
6                              ## [11] [mindepth_statistical]: Min depth for statistical base calling
6                              ## [12] [mindepth_majrule]: Min depth for majority-rule base calling
10000                          ## [13] [maxdepth]: Max cluster depth within samples
0.85                           ## [14] [clust_threshold]: Clustering threshold for de novo assembly
0                              ## [15] [max_barcode_mismatch]: Max number of allowable mismatches in barcodes
2                              ## [16] [filter_adapters]: Filter for adapters/primers (1 or 2=stricter)
35                             ## [17] [filter_min_trim_len]: Min length of reads after adapter trim
2                              ## [18] [max_alleles_consens]: Max alleles per site in consensus sequences
5                              ## [19] [max_Ns_consens]: Max N's (uncalled bases) in consensus (R1, R2)
8                              ## [20] [max_Hs_consens]: Max Hs (heterozygotes) in consensus (R1, R2)
4                              ## [21] [min_samples_locus]: Min # samples per locus for output
20                             ## [22] [max_SNPs_locus]: Max # SNPs per locus (R1, R2)
8                              ## [23] [max_Indels_locus]: Max # of indels per locus (R1, R2)
0.5                            ## [24] [max_shared_Hs_locus]: Max # heterozygous sites per locus (R1, R2)
0, 0, 0, 0                     ## [25] [trim_reads]: Trim raw read edges (R1>, <R1, R2>, <R2) (see docs)
0, 0                           ## [26] [trim_loci]: Trim locus edges (see docs) (R1>, <R1, R2>, <R2)
k                              ## [27] [output_formats]: Output formats (see docs)
                               ## [28] [pop_assign_file]: Path to population assignment file
```

**Comment by laninsky** [comment]
Just found reference to the "-technical-replicate-" method of specifying these guys on gitter and in demultiplex.py (but not on the docs?), so will give that a go and see if I can smush them together that way.

**Comment by laninsky** [comment]
Even using "-technical-replicate-" in the barcodes file, the problem persists:
```
3-technical-replicate-4 TATCTAG
3-technical-replicate-6 TTCGCACA
```
A single file, 3_R1_.fastq.gz, is present in lamprey_technical_fastqs. The s1_demultiplex_stats.txt  has the following entries relevant to sample 3:

```
sample_name                            total_reads
3                                          7239922

sample_name                               true_bar       obs_bar     N_records
3-technical-replicate-4                    TATCTAG       TATCTAG       3008745
3-technical-replicate-6                   TTCGCACA      TTCGCACA       4231177
```
Standard out:
```
 -------------------------------------------------------------
  ipyrad [v.0.7.22]
  Interactive assembly and analysis of RAD-seq data
 -------------------------------------------------------------
  New Assembly: lamprey_technical
  host compute node: [16 cores] on compute-cs-001-p

  Step 1: Demultiplexing fastq data to Samples
^M  [                    ]   0%  chunking large files  | 0:00:00  ^M  [                    ]   0%  chunking large files  | 0:00:00  ^M  [                    ]   0%  
^M  [                    ]   0%  sorting reads         | 0:00:00  ^M  [                    ]   0%  sorting reads         | 0:00:00  ^M  [                    ]   0%  
^M  [                    ]   0%  writing/compressing   | 0:00:00  ^M  [                    ]   0%  writing/compressing   | 0:00:00  ^M  [                    ]   0%  

  Step 2: Filtering reads 
^M  [                    ]   0%  processing reads      | 0:00:01  ^M  [                    ]   0%  processing reads      | 0:00:01  ^M  [                    ]   0%  

  found an error in step2; see ipyrad_log.txt

 Step 3: Clustering/Mapping reads
    Sample not ready for clustering. First run step2 on sample: 3
^M  [                    ]   0%  dereplicating         | 0:00:00  ^M  [                    ]   2%  dereplicating         | 0:00:00  ^M  [##                  ]  13%  
^M  [                    ]   0%  clustering            | 0:00:00  ^M  [                    ]   0%  clustering            | 0:00:00  ^M  [                    ]   0%  
^M  [################### ]  97%  building clusters     | 0:00:00  ^M  [################### ]  97%  building clusters     | 0:00:00  ^M  [################### ]  97%  
^M  [################### ]  97%  chunking              | 0:00:00  ^M  [################### ]  97%  chunking              | 0:00:00  ^M  [################### ]  97%  
```
Log file:
```
 -------------------------------------------------------------
  ipyrad [v.0.7.22]
  Interactive assembly and analysis of RAD-seq data
 -------------------------------------------------------------
  Begin run: 2018-03-22 13:11
  Using args {'preview': False, 'force': False, 'threads': 2, 'results': False, 'quiet': False, 'merge': None, 'ipcluster': 'default', 'cores': 0, 'params': 'params-ipyrad_technical.txt', 'branch': None, 'steps': '1234567', 'debug': False, 'new': None, 'download': None, 'MPI': False}
  Platform info: ('Linux', 'compute-cs-001-p', '2.6.32-279.14.1.el6.x86_64', '#1 SMP Mon Oct 15 13:44:51 EDT 2012', 'x86_64')2018-03-22 13:53:16,254    pid=6623 
       [rawedit.py]    ERROR   error in run_cutadapt(): KeyError('3')
```
All the other samples seem to be progressing through fine (as they did in the last run).

Just in case it is handy, here is Params file:
```
------- ipyrad params file (v.0.7.20)-------------------------------------------
lamprey_technical                        ## [0] [assembly_name]: Assembly name. Used to name output directories for assembly steps
                             ## [1] [project_dir]: Project dir (made in curdir if not present)
SQ0419_CA5EWANXX_s_4_fastq.txt.gz                               ## [2] [raw_fastq_path]: Location of raw non-demultiplexed fastq files
barcodes_technical_reps.txt                               ## [3] [barcodes_path]: Location of barcodes file
                         ## [4] [sorted_fastq_path]: Location of demultiplexed/sorted fastq files
denovo                         ## [5] [assembly_method]: Assembly method (denovo, reference, denovo+reference, denovo-reference)
                               ## [6] [reference_sequence]: Location of reference sequence file
gbs                            ## [7] [datatype]: Datatype (see docs): rad, gbs, ddrad, etc.
TGCAG                         ## [8] [restriction_overhang]: Restriction overhang (cut1,) or (cut1, cut2)
5                              ## [9] [max_low_qual_bases]: Max low quality base calls (Q<20) in a read
33                             ## [10] [phred_Qscore_offset]: phred Q score offset (33 is default and very standard)
6                              ## [11] [mindepth_statistical]: Min depth for statistical base calling
6                              ## [12] [mindepth_majrule]: Min depth for majority-rule base calling
10000                          ## [13] [maxdepth]: Max cluster depth within samples
0.85                           ## [14] [clust_threshold]: Clustering threshold for de novo assembly
0                              ## [15] [max_barcode_mismatch]: Max number of allowable mismatches in barcodes
2                              ## [16] [filter_adapters]: Filter for adapters/primers (1 or 2=stricter)
35                             ## [17] [filter_min_trim_len]: Min length of reads after adapter trim
2                              ## [18] [max_alleles_consens]: Max alleles per site in consensus sequences
5                           ## [19] [max_Ns_consens]: Max N's (uncalled bases) in consensus (R1, R2)
8                           ## [20] [max_Hs_consens]: Max Hs (heterozygotes) in consensus (R1, R2)
4                              ## [21] [min_samples_locus]: Min # samples per locus for output
20                         ## [22] [max_SNPs_locus]: Max # SNPs per locus (R1, R2)
8                           ## [23] [max_Indels_locus]: Max # of indels per locus (R1, R2)
0.5                            ## [24] [max_shared_Hs_locus]: Max # heterozygous sites per locus (R1, R2)
0, 0                     ## [25] [trim_reads]: Trim raw read edges (R1>, <R1, R2>, <R2) (see docs)
0, 0                     ## [26] [trim_loci]: Trim locus edges (see docs) (R1>, <R1, R2>, <R2)
k                        ## [27] [output_formats]: Output formats (see docs)
                               ## [28] [pop_assign_file]: Path to population assignment file
```
Again, any advice would be greatly appreciated.

**Comment by isaacovercast** [comment]
Thanks for all the great detail! This is an ideal ticket, from the perspective of information content.

In terms of the problem, that is weird, not sure what's going on. Do other samples have single digit names like this one? The naming scheme for this sample might be an edge case. I've never seen this before, and it _should_ work, but these kinds of things have a way of not working sometimes. If other samples have single digit names, then this is certainly not the problem, and you're probably right it probably has to do with replicates with different barcode lengths. Can you re-run step 2 with this sample include and add the `-d` flag? This will generate more verbose output in the log file. Paste anything that looks relevant in this ticket.

**Comment by laninsky** [comment]
Hi Isaac - I did try renaming that guy to 3.0 rather than 3 and rerunning ipyrad a couple of days ago, just in case that was the issue, but that sample still failed at the same place/same errors. There are a number of other samples which are named as [0-9].[0-9] that got through fine (e.g. 1.6, 8.1), so it doesn't seem like it is the name per se. I'll get on rerunning that step with the -d option and shooting you guys the output. Thanks for getting back to me!

**Comment by isaacovercast** [comment]
Ok, good, I kind of figured it wasn't the naming scheme, but wanted to be
sure. Let me know if you see anything in the log file. Also, if you want to
dropbox me the 3.fastq.gz file from the _fastqs directory I can take a look
at it on my end.

**Comment by laninsky** [comment]
Hi Isaac - I don't think there is anything else to go on here (I think I probably actually used the -d command earlier too to try and see if I could figure out what was going on myself). Attached the relevant bits here just in case.

The command:
```
export PATH="/projects/uoo00110/bin/miniconda2/bin:$PATH"
export HOME="/projects/uoo00110"
ipcluster start --n 16 --daemonize
sleep 240
ipyrad -p params-lamprey_3_only.txt -s 2 -f -d --ipcluster
```
The log: 
```
Begin run: 2018-03-28 11:04
  Using args {'preview': False, 'force': True, 'threads': 2, 'results': False, 'quiet': False, 'merge': None, 'ipcluster': 'default', 'cores': 0, 'params': 'params-lamprey_3_only.txt', 'branch': None, 'steps': '2', 'debug': True, 'new': None, 'download': None, 'MPI': False}
  Platform info: ('Linux', 'compute-e1-054-p', '2.6.32-279.14.1.el6.x86_64', '#1 SMP Mon Oct 15 13:44:51 EDT 2012', 'x86_64')2018-03-28 11:04:16,973    pid=6988        [load.py]       DEBUG   skipping: no svd results present in old assembly
2018-03-28 11:04:26,185         pid=6988        [rawedit.py]    INFO    [7239922.0]
2018-03-28 11:04:32,478         pid=6988        [rawedit.py]    ERROR   error in run_cutadapt(): KeyError('3')
2018-03-28 11:04:34,550         pid=6988        [__init__.py]   INFO    debugging turned off
```
Relevant bits from standard out:
```
/projects/uoo00110/bin/miniconda2/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
INFO:ipyrad.assemble.rawedit:[7239922.0]

ipyrad [v.0.7.22]
  Interactive assembly and analysis of RAD-seq data
 -------------------------------------------------------------
  loading Assembly: lamprey_3_only
  from saved path: /gpfs1m~/lamprey_3_only.json
  Warning: technical replicates (same name) will be combined.
  Warning: technical replicates (same name) will be combined.
  host compute node: [16 cores] on compute-e1-054-p

#] 100%  processing reads      | 0:00:06  ERROR:ipyrad.assemble.rawedit:error in run_cutadapt(): KeyError('3')

  found an error in step2; see ipyrad_log.txt
INFO:ipyrad:debugging turned off
```
Relevant bits from barcode file:
```
3       TATCTAG
3       TTCGCACA
```
Params file:
```
------- ipyrad params file (v.0.7.22)-------------------------------------------
lamprey_3_only                 ## [0] [assembly_name]: Assembly name. Used to name output directories for assembly steps
/gpfs1m/projects/uoo00110      ## [1] [project_dir]: Project dir (made in curdir if not present)
/gpfs1m/projects/uoo00110/SQ0419_CA5EWANXX_s_4_fastq.txt.gz ## [2] [raw_fastq_path]: Location of raw non-demultiplexed fastq files
/gpfs1m/projects/uoo00110/barcodes_3merged.txt ## [3] [barcodes_path]: Location of barcodes file
/projects/uoo00110/lamprey_fastqs/*.fastq.gz                               ## [4] [sorted_fastq_path]: Location of demultiplexed/sorted fastq files
denovo                         ## [5] [assembly_method]: Assembly method (denovo, reference, denovo+reference, denovo-reference)
                               ## [6] [reference_sequence]: Location of reference sequence file
gbs                            ## [7] [datatype]: Datatype (see docs): rad, gbs, ddrad, etc.
TGCAG, TGCAG                   ## [8] [restriction_overhang]: Restriction overhang (cut1,) or (cut1, cut2)
5                              ## [9] [max_low_qual_bases]: Max low quality base calls (Q<20) in a read
33                             ## [10] [phred_Qscore_offset]: phred Q score offset (33 is default and very standard)
6                              ## [11] [mindepth_statistical]: Min depth for statistical base calling
6                              ## [12] [mindepth_majrule]: Min depth for majority-rule base calling
10000                          ## [13] [maxdepth]: Max cluster depth within samples
0.85                           ## [14] [clust_threshold]: Clustering threshold for de novo assembly
0                              ## [15] [max_barcode_mismatch]: Max number of allowable mismatches in barcodes
2                              ## [16] [filter_adapters]: Filter for adapters/primers (1 or 2=stricter)
35                             ## [17] [filter_min_trim_len]: Min length of reads after adapter trim
2                              ## [18] [max_alleles_consens]: Max alleles per site in consensus sequences
5                              ## [19] [max_Ns_consens]: Max N's (uncalled bases) in consensus (R1, R2)
8                              ## [20] [max_Hs_consens]: Max Hs (heterozygotes) in consensus (R1, R2)
4                              ## [21] [min_samples_locus]: Min # samples per locus for output
20                             ## [22] [max_SNPs_locus]: Max # SNPs per locus (R1, R2)
8                              ## [23] [max_Indels_locus]: Max # of indels per locus (R1, R2)
0.5                            ## [24] [max_shared_Hs_locus]: Max # heterozygous sites per locus (R1, R2)
0, 0, 0, 0                     ## [25] [trim_reads]: Trim raw read edges (R1>, <R1, R2>, <R2) (see docs)
0, 0                           ## [26] [trim_loci]: Trim locus edges (see docs) (R1>, <R1, R2>, <R2)
k                              ## [27] [output_formats]: Output formats (see docs)
                               ## [28] [pop_assign_file]: Path to population assignment file
```

**Comment by isaacovercast** [comment]
In data.barcodes i see:
```
        "barcodes":{
            "3-technical-replicate-6":"TTCGCACA",
            "3-technical-replicate-4":"TATCTAG"
```
But step 1 strips the -technical-replicate-* part for the sample name, so the sample name in the assembly is 3.
Then during step 2 when constructing the command for cutadapt it tries this:
```
        fullcomp(data.barcodes[sample.name])[::-1]
```
but here sample.name == 3, and this is not in the barcodes in the assembly....
Step 1 works great otherwise (all reads from all technical replicates are gathered into the 3 sample).

**Comment by isaacovercast** [comment]
This is only really a problem for SE `gbs`.

**Comment by isaacovercast** [comment]
Ok, so this is only really happening because of a confluence of the data
being single end gbs and also this sample being a technical replicate.
There are two ways to fix this, the easy way and the "right" way:

The "easy" way: in your params file set the `datatype` to rad. Run step 2.
Set the datatype back to gbs before running step 3. The outcome of this is
that the data will not be filtered for barcodes on the 3' end in case the
read overruns the length of the fragment. If you have short fragments
(which you do) then you'll end up with some barcodes being treated as
"real" data. The up side is it really shouldn't matter that much because
the barcodes should be invariant.

The "right" way: Run each sample through step 2 independently as the gbs
datatype. Go into the *_edits directory and take the trimmed reads from one
of the replicates and concatenate them on to the end of the other trimmed
reads file, then branch and remove the sample that donated reads. This is
more right in that it'll filter the barcodes properly, but it is also more
annoying and requires a bit more comfort "hacking" on the data. Also, the
count of raw reads for the merged sample will be wrong.

Either one of those should work. Let me know which you prefer and if you
have any luck.

**Comment by laninsky** [comment]
Hi Isaac - sorry for going AWOL: ended up getting busy with other projects. I went with the "right" way, and the run finished successfully, even for our 'problem child'. I'm pasting in the commands here just in case it helps someone else with the same issue, and thanks again for being such an awesome troubleshooting resource!
```
#Running steps 1-2 on the original data (needed the force option because it "remembered" one of my previous attempts)
export PATH="/projects/uoo00110/bin/miniconda2/bin:$PATH"
export HOME="/projects/uoo00110"
ipcluster start --n 16 --daemonize
sleep 240
ipyrad -p params-lamprey.txt -s 12 -f --ipcluster

# Catting 3.6 on to the end of 3.4
cd lamprey_edits
cp 3.4.trimmed_R1_.fastq.gz 3.4.trimmed_R1_.fastq.gz_backup
cat 3.6.trimmed_R1_.fastq.gz >> 3.4.trimmed_R1_.fastq.gz

## Note, inadvertently using zcat without following gzip -c command e.g.
#zcat 3.6.trimmed_R1_.fastq.gz >> 3.4.trimmed_R1_.fastq.gz
## Led to the following error in step 3
#Reading file /dev/stdin
#
#Fatal error: Invalid line 11952180 in FASTQ file: Illegal character ' '
#)

cd ..

# branching the assembly and removing 3.6 (which is now merged with 3.4)
ipyrad -p params-lamprey.txt -b 3_6_merged  - 3.6

#Running ipyrad on the "new" version
ipyrad -p params-3_6_merged.txt -s 34567 --ipcluster
```

**Comment by isaacovercast** [comment]
Great! Thanks for letting us know that it worked.
